### PR TITLE
JVM: be more careful when removing unused constants 

### PR DESCRIPTION
**Pull request by pyos**
1. if an argument of a `pop` cannot be removed, then all other potential arguments of that `pop` can't be removed either, and the same applies to other `pop`s that touch them;
2. the same is true for primitive conversions, but this is even trickier to implement correctly, so I simply did the same thing as with boxing operators: replace the conversion itself with a `pop` and keep the argument as-is.

Somehow this actually removes *more* redundant primitive type conversions than the old code in a couple bytecode text tests, so I've patched them to kind of use the value, forcing the instructions to stay.

 #KT-46921 Fixed

### CODE REVIEW

File: compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/boxing/PopBackwardPropagationTransformer.kt [hunk 1]
```diff
@@ -32,91 +29,66 @@ import org.jetbrains.org.objectweb.asm.tree.analysis.SourceInterpreter
...
         private val insns = insnList.toArray()
 
         private val dontTouchInsnIndices = BitSet(insns.size)
-        private val transformations = hashMapOf<AbstractInsnNode, Transformation>()
-        private val removableNops = hashSetOf<InsnNode>()
-
-        private val frames by lazy { analyzeMethodBody() }
 
         fun transform() {
-            if (insns.none { it.isPop() || it.isPurePush() }) return
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/boxing/PopBackwardPropagationTransformer.kt:70]: Why did you remove this line? Looks like it was added intentionally to avoid calling `analyze` (a5d6559056ade2942dacd70d070d94cbd82c3f47, 9356c1e0ef67cd7064b5e6bcc5fbe786128384cf)

Comment (udalov) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/boxing/PopBackwardPropagationTransformer.kt:70]: Oops, just noticed that it was moved above, sorry!

File: compiler/testData/codegen/box/optimizations/kt46921.kt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// TARGET_PLATFORM: JVM
```
Comment (udalov) [inline compiler/testData/codegen/box/optimizations/kt46921.kt:1]: `TARGET_BACKEND` + regenerate tests pls.

But also consider moving this test to `boxInline`, to test the two-module scenario too

Comment (pyos) [reply compiler/testData/codegen/box/optimizations/kt46921.kt]: I'm not sure it makes sense to test this as multiple modules since the purpose of `inline` here is to generate a specific bytecode pattern, not to test the inliner.

**Comment by udalov** [comment]
Thank you!
